cluster/assemblestate/assemblestate.go

```diff
@@ -0,0 +1,880 @@
+	// their provenance is this local node, since we don't persist which routes
+	// came from which peer. this will lead to our local node doing some wasted
+	// publications, but that is okay.
+	if _, _, err := sel.RecordRoutes(config.RDT, session.Routes); err != nil {
+		return nil, err
+	}
+
+	// for any peers that we already trust and know their address, we can safely
+	// inform the selector that the route from our local node to that peer can
+	// be published
+	for fp, peer := range validated.trusted {
+		addr, ok := validated.addresses[fp]
+		if !ok {
+			continue
+		}
+
+		sel.AddAuthoritativeRoute(peer.RDT, addr)
+	}
+
+	// calculate the HMAC once for this device's authentication
+	authHMAC := CalculateHMAC(config.RDT, CalculateFP(cert.Certificate[0]), config.Secret)
+
+	as := AssembleState{
+		initiated:    session.Initiated,
+		config:       config,
+		logger:       log,
+		commit:       commit,
+		clock:        config.Clock,
+		cert:         cert,
+		authHMAC:     authHMAC,
+		trusted:      validated.trusted,
+		fingerprints: validated.fingerprints,
+		addresses:    validated.addresses,
+		discovered:   validated.discovered,
+		devices:      devices,
+		selector:     sel,
+	}
+
+	return &as, nil
+}
+
+type AssembleConfig struct {
+	Secret  string           `json:"secret"`
+	RDT     DeviceToken      `json:"rdt"`
+	IP      net.IP           `json:"ip"`
+	Port    int              `json:"port"`
+	TLSCert []byte           `json:"cert"`
+	TLSKey  []byte           `json:"key"`
+	Clock   func() time.Time `json:"-"`
+}
+
+// publishAuth uses the given [Client] to publish to each device. If publication
+// succeeds, then the certificate returned by [Client.Untrusted] (the
+// certificate that the peer used to communicate with us) is associated with the
+// address that we reached them at.
+//
+// If the certificate the peer used is already trusted (they've already
+// published their auth message to us), then we verify the route from this local
+// node to that peer.
+func (as *AssembleState) publishAuth(ctx context.Context, addresses []string, client Client) error {
+	as.lock.Lock()
+	defer as.lock.Unlock()
+
+	for _, addr := range addresses {
+		if as.discovered[addr] {
+			continue
+		}
+
+		as.lock.Unlock()
+		cert, err := client.Untrusted(ctx, addr, "auth", Auth{
+			HMAC: as.authHMAC,
+			RDT:  as.config.RDT,
+		})
+		as.lock.Lock()
+
+		if err != nil {
+			as.debugf("cannot send message to peer: %v", err)
+			continue
+		}
+
+		as.debugf("sent auth message to %s", addr)
+
+		fp := CalculateFP(cert)
+
+		if expected, ok := as.addresses[fp]; ok && expected != addr {
+			return fmt.Errorf("found new address %s using same certificate as other address %s", addr, expected)
+		}
+
+		// TODO: consider devices with multiple addresses
+		as.addresses[fp] = addr
+		as.discovered[addr] = true
+
+		if p, ok := as.trusted[fp]; ok {
+			as.selector.AddAuthoritativeRoute(p.RDT, addr)
+		}
+	}
+
+	as.commit(as.export())
```

Comment:
it's a bit annoying that is hard to tell from the method names what commits and what not